src/compas/datastructures/tree/tree.py

```diff
@@ -0,0 +1,259 @@
+    >>> leaf1 = TreeNode('leaf1')
+    >>> leaf2 = TreeNode('leaf2')
+    >>> tree.add_root(root)
+    >>> root.add(branch)
+    >>> branch.add(leaf1)
+    >>> branch.add(leaf2)
+    >>> print(tree)
+    <Tree with 4 nodes, 1 branches, and 2 leaves>
+    >>> tree.print()
+    <TreeNode root>
+        <TreeNode branch>
+            <TreeNode leaf2>
+            <TreeNode leaf1>
+
+    """
+
+    JSONSCHEMA = {}
+
+    def __init__(self, name: str = None, attributes: dict = None):
+        super(Tree, self).__init__()
+        self.name = name
+        self.attributes = attributes or {}
+        self._root = None
+        self._nodes = set()
+
+    @property
+    def data(self):
+        def get_node_data(node):
+            return {
+                "name": node.name,
+                "attributes": node.attributes,
+                "children": [get_node_data(child) for child in node.children],
+            }
+
+        return {
+            "name": self.name,
+            "root": get_node_data(self.root),
+            "attributes": self.attributes,
+        }
+
+    @data.setter
+    def data(self, data):
+        self.name = data["name"]
+        self.attributes = data["attributes"]
+
+        def node_from_data(data):
+            node = TreeNode(data["name"], data["attributes"])
+            for child in data["children"]:
+                node.add(node_from_data(child))
+            return node
+
+        self.add_root(node_from_data(data["root"]))
+
+    @property
+    def root(self):
+        return self._root
+
+    def add_root(self, node: TreeNode):
+        """Add a root node to the tree."""
+        if not isinstance(node, TreeNode):
+            raise TypeError("The node is not a TreeNode object.")
+        if not node.is_root:
+            raise ValueError("The node is already part of another tree.")
+        self._root = node
+        node._tree = self
+        self._nodes.add(node)
+
+    def add(self, node: TreeNode, parent: TreeNode):
+        """Add a node to the tree."""
+        if not isinstance(node, TreeNode):
+            raise TypeError("The node is not a TreeNode object.")
+        if self.root is None:
+            raise ValueError("The tree has no root node, use add_root() first.")
+        else:
+            parent.add(node)
+
+    @property
+    def nodes(self):
+        return self._nodes
+
+    def remove_root(self):
+        """Remove the root node from the tree."""
+        self._root._tree = None
+        self.nodes.remove(self._root)
+        self._root = None
+
+    def remove(self, node: TreeNode):
+        """Remove a node from the tree."""
+        if node == self.root:
+            self.remove_root()
+        else:
+            node.parent.remove(node)
+
+    @property
+    def leaves(self):
+        for node in self.nodes:
+            if node.is_leaf:
+                yield node
+
+    def __repr__(self):
+        nodes = len(self.nodes)
+        branches = sum(1 for node in self.nodes if node.is_branch)
```

Comment:
This feels like an extremely expensive operation, for a statement that will be triggered inadvertently without considering its cost. I would remove all the summing of branches and leaves.